### PR TITLE
usm: http2: Split eBPF telemetry

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Splits the eBPF telemetry map into a plaintext telemetry and TLS telemetry maps, by creating a new map in the same format.
The PR modifies the user mode code to collect the telemetry from the 2 maps, and unify them into the same metric, with a dedicated different tag (`encrypted:true` or `encrypted:false`).

### Motivation

Our plaintext and TLS http2 program similar but not identical. The programs can handle different use-cases (for example, the TLS version often sees in a single call only the frame header, and in the following calls gets the frame payload, while in the plaintext version, the network version, we often get the entire payload, or a randomly cut version of it, due to TCP split). Also, the program has different limitations (plaintext can process up to 240 frame, while TLS can process up to 120). To ensure our metrics are not polluted, we needed to split them and to have the ability to know which program contributed to a certain metric.

### Additional Notes
Additional cost of 4K to the kernel memory (the minimal cost of a map).

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/protocols/http2/protocol_testutils.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+package http2
+
+import (
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// GetHTTP2KernelTelemetry returns the HTTP2 kernel telemetry
+func (p *Protocol) GetHTTP2KernelTelemetry() (*HTTP2Telemetry, error) {
```
Comment (amitslavin) [inline pkg/network/protocols/http2/protocol_testutils.go]: Why do you think it is better over a utility test only?

File: pkg/network/protocols/http2/protocol.go [hunk 2]
```diff
@@ -265,17 +266,30 @@ func (p *Protocol) PostStart(mgr *manager.Manager) error {
...
+	mp, err := getMap(mgr, telemetryMap)
 	if err != nil {
-		log.Warnf("unable to get http2 telemetry map: %s", err)
+		log.Warn(err)
 		return
 	}
 
-	if mp == nil {
-		log.Warn("http2 telemetry map is nil")
+	tlsMap, err := getMap(mgr, tlsTelemetryMap)
```
Comment (amitslavin) [inline pkg/network/protocols/http2/protocol.go:287]: shouldn't we perform it only if tls is enabled? I mean do we want that each time we will try to get the tls map as well to the plain map?

File: pkg/network/protocols/http2/protocol.go [hunk 3]
```diff
@@ -265,17 +266,30 @@ func (p *Protocol) PostStart(mgr *manager.Manager) error {
...
+	if err != nil {
```
Comment (amitslavin) [inline pkg/network/protocols/http2/protocol.go:288]: maybe lets wrap the error to be able to easily understand if the issue is realted to the plain or not

File: pkg/network/protocols/http2/protocol.go [hunk 4]
```diff
@@ -265,17 +266,30 @@ func (p *Protocol) PostStart(mgr *manager.Manager) error {
...
+	}
+	return m, nil
+}
+
 func (p *Protocol) updateKernelTelemetry(mgr *manager.Manager) {
-	mp, _, err := mgr.GetMap(telemetryMap)
+	mp, err := getMap(mgr, telemetryMap)
 	if err != nil {
-		log.Warnf("unable to get http2 telemetry map: %s", err)
+		log.Warn(err)
```
Comment (amitslavin) [inline pkg/network/protocols/http2/protocol.go:283]: maybe lets wrap the error to be able to easily understand if the issue is realted to the plain or not

File: pkg/network/protocols/http2/protocol.go [hunk 5]
```diff
@@ -265,17 +266,30 @@ func (p *Protocol) PostStart(mgr *manager.Manager) error {
...
 		return
 	}
 
-	if mp == nil {
```
Comment (amitslavin) [inline pkg/network/protocols/http2/protocol.go:275]: you removed this check in purpose? the map received can be nil which would be a problem

**Review by amitslavin — CHANGES_REQUESTED** [review]

**Comment by amitslavin** [comment]
Did you had a chance to test it? (not with ut's) to actually see the metrics? (maybe with load test?)

File: pkg/network/protocols/http2/protocol_testutils.go (hunk 1 above)
Comment (guyarb) [reply pkg/network/protocols/http2/protocol_testutils.go]: This is a good question.
So originally we added this as the tests using it was in the `tracer` package, and we didn't have access to USM's manager.
But now, as we are in the usm package, we are no longer limited, so I'll move it to a test utility indeed.

File: pkg/network/protocols/http2/protocol.go (hunk 2 above)
Comment (guyarb) [reply pkg/network/protocols/http2/protocol.go:287]: this is a good question, the protocol does not know if tls is enabled, this is in our todo list.
For example today we're loading all probes and maps (including TLS probes and maps) if http2 is enabled, regardless of the state of any of the TLS capabilities.
I'd suggest to postpone the change you're suggesting to the PR that will make the protocols to be aware of the TLS state.

File: pkg/network/protocols/http2/protocol.go (hunk 3 above)
Comment (guyarb) [reply pkg/network/protocols/http2/protocol.go:288]: we do know, as the inner error contains the map name

File: pkg/network/protocols/http2/protocol.go (hunk 4 above)
Comment (guyarb) [reply pkg/network/protocols/http2/protocol.go:283]: https://github.com/DataDog/datadog-agent/pull/22460#discussion_r1472945558

File: pkg/network/protocols/http2/protocol.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/protocols/http2/protocol.go:275]: it was moved to `getMap` if `mp` is nil, then an error will be returned instead

**Comment by guyarb** [comment]
> Did you had a chance to test it? (not with ut's) to actually see the metrics? (maybe with load test?)

I waited for the CI to build an image :) 
I'm planning to dogfooding it before merging

File: pkg/network/protocols/http2/protocol.go (hunk 2 above)
Comment (amitslavin) [reply pkg/network/protocols/http2/protocol.go:287]: sounds good as long we it is in our todo list :)

**Review by amitslavin — APPROVED** [review]

**Pull request merged**